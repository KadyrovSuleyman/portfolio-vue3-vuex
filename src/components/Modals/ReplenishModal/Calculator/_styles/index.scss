@forward '__layout';
@forward '__font';
@forward '__color';
@forward '__cursor';
