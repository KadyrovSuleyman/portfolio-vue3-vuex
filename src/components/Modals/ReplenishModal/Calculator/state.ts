import { Store } from 'vuex';

export interface StateT {
  value: string,
  setValue: (input: string) => void,

  maxValue: number,
  replenish: (input: string) => Promise<any>,
}

const adaptState = (store: Store<any>, props: any): StateT => ({
  maxValue: store.getters['stake/replenishMax'],
  replenish: (input: string) => store.dispatch('waiting/replenish', Number(input)),

  value: props.value,
  setValue: props.setValue,
});

export default adaptState;
