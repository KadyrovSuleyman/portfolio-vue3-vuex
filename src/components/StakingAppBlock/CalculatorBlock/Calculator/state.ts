import { Store } from 'vuex';

export interface StateT {
  maxValue: number,
  minValue: number,
  period: number,
  apy: number,

  disabled: boolean,

  text: string,
  setText: (newInput: string) => void,

  reward: number,

  submit: () => Promise<any>,
}

const adaptState = (store: Store<any>): StateT => ({
  maxValue: store.getters['tariff/amountMax'],
  minValue: store.getters['tariff/amountMin'],
  period: store.getters['tariff/period'],
  apy: store.getters['tariff/apy'],

  disabled: !store.getters['stake/isAllowed'],

  text: store.state.stake.inputValue,
  setText: (newInput: string) => store.commit('stake/input', newInput),

  reward: store.getters['stake/reward'],

  submit: () => store.dispatch('waiting/stake'),
});

export default adaptState;
