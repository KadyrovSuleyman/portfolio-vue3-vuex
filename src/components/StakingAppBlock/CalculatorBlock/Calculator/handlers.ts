import { ComputedRef } from 'vue';
import { isValidInput } from './logic';
import { StateT } from './state';

export const createMaxButtonClickHandler = (
  state: ComputedRef<StateT>,
  inputRef: HTMLInputElement,
) => () => {
  state.value.setText(String(state.value.maxValue));
  inputRef.focus();
};

export const createOnInputHandler = (
  state: ComputedRef<StateT>,
) => (payload: KeyboardEvent) => {
  const target = (payload.target as HTMLInputElement);

  if (!isValidInput(target.value)) {
    target.value = state.value.text;
    return;
  }

  if (Number(target.value) > state.value.maxValue) {
    target.value = String(state.value.maxValue);
  }

  state.value.setText(target.value);
};

export const createOnKeyupHandler = (
  state: ComputedRef<StateT>,
) => (payload: KeyboardEvent) => {
  if (payload.key === 'Enter') {
    state.value.submit();
  }
};
