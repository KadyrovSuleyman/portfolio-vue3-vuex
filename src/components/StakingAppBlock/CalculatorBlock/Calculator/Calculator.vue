<script setup lang="ts">

import { computed, onMounted, ref } from 'vue';
import Div from '@/elements/Div/Div.vue';
import propsObj from '@/elements/propsObj';
import Button from '@/elements/Button/Button.vue';
import Span from '@/elements/Span/Span.vue';
import { useStore } from 'vuex';
import PromtInput from './PromtInput/PromtInput.vue';
import { validate, correctState } from './logic';
import adaptState from './state';
import { inputValueWatcher } from './watchers';
import { createMaxButtonClickHandler, createOnInputHandler, createOnKeyupHandler } from './handlers';

// eslint-disable-next-line no-undef
const props = defineProps({ ...propsObj });
const comp = computed(() => ({ elem: props.elem || 'calculator' }));

const store = useStore();
const state = computed(() => adaptState(store));

const errorText = computed(() => validate(state.value.text, state.value.minValue));
const correct = computed(() => correctState(state.value.text, state.value.minValue));

const promtInput = ref(null) as any;

const onButtonClick = ref<() => void>();
onMounted(() => {
  onButtonClick.value = createMaxButtonClickHandler(state, promtInput.value.input.element);
});
const onInput = createOnInputHandler(state);
const onKeyup = createOnKeyupHandler(state);

inputValueWatcher(state);

</script>

<template>
  <Div :block="props.block" :elem="comp.elem" :mods="props.mods">
    <PromtInput :block="comp.elem" :promtText="errorText" :correct="correct" ref="promtInput"
      :text="state.text" :onInput="onInput" :disabled="state.disabled" :onKeyup="onKeyup"
    />
    <Button :block="comp.elem" :onClick="onButtonClick" :disabled="state.disabled">
      Max
    </Button>

    <Div :block="comp.elem">
      <Span :block="comp.elem" :elem="'periodSpan'">
        Reward {{ state.period ? `for ${state.period} Days` : ''}}:
      </Span>
      <Span v-if="correct === 'true'" :block="comp.elem" :elem="'rewardSpan'">
        {{ state.reward }} TKN
      </Span>
    </Div>
  </Div>
</template>

<style lang="scss">
  @use 'styles' as *;
</style>
