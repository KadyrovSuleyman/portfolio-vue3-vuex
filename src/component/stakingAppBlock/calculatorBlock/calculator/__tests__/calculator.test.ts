import { mount, VueWrapper } from '@vue/test-utils';
import { createStore, Store } from 'vuex';
import Calculator from '../calculator.vue';

jest.mock('../adapter');
jest.mock('../logic');

let wrapper: VueWrapper<any>;
afterEach(() => {
  wrapper.unmount();
});

let store: Store<any>;
beforeEach(() => {
  store = createStore<any>({
    state: {
      maxValue: 1000,
      minValue: 500,
      period: 30,
      apy: 150,

      text: '',
      setText: (newInput: string) => store.commit('change', {
        text: newInput,
      }),

      reward: 0,
    },
    mutations: {
      change: (state, obj: { [name: string]: any }) => {
        Object.keys(obj).forEach((index) => {
          state[index] = obj[index];
        });
      },
    },
  });
});

it('calculator renders', () => {
  wrapper = mount(Calculator, {
    global: { plugins: [store] },
  });

  expect(wrapper.find('.calculator').classes()).toEqual(['calculator']);

  expect(wrapper.find('.calculator-promtInput').exists()).toBeTruthy();
  expect(wrapper.find('.calculator-button').exists()).toBeTruthy();
  expect(wrapper.find('.calculator-div').exists()).toBeTruthy();
  expect(wrapper.find('.calculator-periodSpan').exists()).toBeTruthy();
  expect(wrapper.find('.calculator-rewardSpan').exists()).toBeFalsy();
});

it('watchs props changes', async () => {
  wrapper = mount(Calculator, {
    global: { plugins: [store] },
  });
  expect(wrapper.find('.calculator').classes()).toEqual(['calculator']);

  await wrapper.setProps({
    mods: {
      selected: true,
    },
  });
  expect(wrapper.find('.calculator').classes()).toEqual(['calculator', 'calculator__selected']);

  await wrapper.setProps({
    mods: {
      selected: false,
    },
  });
  expect(wrapper.find('.calculator').classes()).toEqual(['calculator']);
});

describe('watching outer store', () => {
  it('maxValue', async () => {
    wrapper = mount(Calculator, {
      global: { plugins: [store] },
    });
    await wrapper.vm.$nextTick();
    expect(wrapper.find('.calculator').exists()).toBeTruthy();

    const input = wrapper.find('.promtInput-input');
    expect((input.element as HTMLInputElement).value).toBe('');

    const max = wrapper.find('.calculator-button');
    await max.trigger('click');
    expect((input.element as HTMLInputElement).value).toBe(`${store.state.maxValue}`);

    ((input.element as HTMLInputElement).value) = String(500);
    await input.trigger('input');
    expect((input.element as HTMLInputElement).value).toBe('500');

    store.commit('change', {
      maxValue: 800,
    });
    await max.trigger('click');
    expect(store.state.maxValue).toBe(800);
    expect((input.element as HTMLInputElement).value).toBe('800');
  });

  it('period', async () => {
    wrapper = mount(Calculator, {
      global: { plugins: [store] },
    });
    expect(wrapper.find('.calculator').exists()).toBeTruthy();
    expect(wrapper.find('.calculator-periodSpan').text())
      .toBe(`Reward for ${store.state.period} Days:`);

    store.commit('change', {
      period: 90,
    });
    await wrapper.vm.$nextTick();
    expect(wrapper.find('.calculator-periodSpan').text())
      .toBe('Reward for 90 Days:');
  });
});
