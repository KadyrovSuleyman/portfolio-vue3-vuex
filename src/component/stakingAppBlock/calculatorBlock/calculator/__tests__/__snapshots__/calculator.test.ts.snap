// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calculator renders 1`] = `
<div class="calculator">
  <div class="calculator-promtInput"><input class="promtInput-input">
    <!--v-if-->
  </div><button class="calculator-button"> Max </button>
  <div class="calculator-div"><span class="calculator-periodSpan"> Reward for 30 Days: </span>
    <!--v-if-->
  </div>
</div>
`;
